State: Use go-datastore to implement the state interface

hsanjuan opened the pull request:
Since the beginning, we have used a Go map to store the shared state (pinset)
in memory. The mapstate knew how to serialize itself so that libp2p-raft would
know how to write to disk when it:

* Saved snapshots of the state on shutdown
* Sent the state to a newcomer peer

hashicorp.Raft assumes an in-memory state which is snapshotted from time to
time and read from disk on boot.

This commit adds a `dsstate` implementation of the state interface using
`go-datastore`. This allows to effortlessly switch to a disk-backed state in
the future (as we will need), and also have at our disposal the different
implementations and utilities of Datastore for fine-tuning (caching, batching
etc.).

`mapstate` has been reworked to use dsstate. Ideally, we would not even need
`mapstate`, as it would suffice to initialize `dsstate` with a
`MapDatastore`. BUT, we still need it separate to be able to auto-migrate to
the new format.

This will be the last migration with the current system. Once this has been
released and users have been able to upgrade we will just remove `mapstate` as
it is now.

License: MIT
Signed-off-by: Hector Sanjuan <code@hector.link>

Depends on https://github.com/libp2p/go-libp2p-raft/pull/50

hsanjuan commented:
Also, kind of depends on discussion #654 . If we slightly changed how things are  serialized soon, I'd rather not have to migrate right after this.

lanzafame commented:
LGTM. Only thing would be a better explanation of what the migration process is and how the migration process will be different in the future, i.e. post v6.

lanzafame commented:
@hsanjuan It seems that in some cases, this PR has slowed down cluster as there have now been 2 test timeout failures. I am happy to bump the timeout up again and then chip away at getting it down again.

hsanjuan commented:
> Also, kind of depends on discussion #654 . If we slightly changed how things are serialized soon, I'd rather not have to migrate right after this.

I don't fully know. But since checking state version will mean opening the datastore and reading one key, we know that will be different. The migration will be looping the keys and overwriting them with their updated form.

The key idea is that Marshal/Unmarshal will either not be used, or can happen normally regardless of the state version (unlike now).

hsanjuan commented:
> @hsanjuan It seems that in some cases, this PR has slowed down cluster as there have now been 2 test timeout failures. I am happy to bump the timeout up again and then chip away at getting it down again.

I only see jenkins, which I retriggered now. It make take a while before I merge this anyway.

Does this approval include changes in go-libp2p-raft ?

lanzafame commented:
@hsanjuan I just added some comments on the go-libp2p-raft PR.

hsanjuan commented:
@lanzafame do me a favour and have a look to the last commits (added tracing) and lmk if we can leave this ready for merge, as I want it to go before #670 .

lanzafame commented on cluster.go:
```diff
@@ -739,20 +739,7 @@ func (c *Cluster) Join(ctx context.Context, addr ma.Multiaddr) error {
-		case <-dhtBstCtx.Done(): // shut down the process
-			proc.Close()
-		}
+		c.dht.BootstrapOnce(c.ctx, dht.DefaultBootstrapConfig)
```
```suggestion
		c.dht.BootstrapOnce(ctx, dht.DefaultBootstrapConfig)
```
`ctx` has already been derived from the `c.ctx` at the top of the function.

lanzafame commented on api/types.go:
```diff
@@ -704,10 +707,11 @@ func (pT PinType) String() string {
-	ReplicationFactorMax int    `json:"replication_factor_max"`
-	Name                 string `json:"name"`
-	ShardSize            uint64 `json:"shard_size"`
+	ReplicationFactorMin int               `json:"replication_factor_min" codec:"rn,omitempty"`
```
@hsanjuan Is there a compile-time or test-time check that we can add so that we don't conflict with the codec variable?

hsanjuan replied:
As discussed, tests break in most cases when screwing up the codec field names.

hsanjuan replied:
Additionally I have added a test that verifies that there are no dup fields.

GitCop commented:
There were the following issues with your Pull Request

 * Commit: 1491a01ec8028fd4e593bcff6a33cf66cce6d060
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

GitCop commented:
There were the following issues with your Pull Request

 * Commit: 1491a01ec8028fd4e593bcff6a33cf66cce6d060
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

hsanjuan commented on api/pb/types.proto:
```diff
@@ -0,0 +1,23 @@
+
+message Pin {
+  bytes Cid = 1;
+  enum PinType {
```
I can't have an enum called `Type` and a field called `Type`

hsanjuan replied:
let's leave this as is, I don't mind so much and we can change it later.

hsanjuan commented:
@lanzafame I squashed, rebased and cleaned up. Then I tested with larger states and had to fix some issues (latest commit). Can you check again? I think we can merge then.

lanzafame commented:
@hsanjuan LGTM
